### DISCUSSION TITLE
test_competitor package as a reusable and extensible component for medium level task coordination

### DISCUSSION THREAD

**Discussion by AravindaDP** [usnistgov/ARIAC#246] [Ideas]
While test_competitor package is provided as only a reference implementation to demonstrate how a sample competitor package can be created, in my opinion it also has a potential to be included as part of competitors solution to handle medium level task handling if applicable. Idea is to use test_competitor functions for medium level tasks like pick part in x location to y location etc. and then build a higher-level agility framework that handles task coordination between robots, handle agility challenges etc. (Of course, competitors are free to build up from ground up for more flexibility of overall architecture)

With this aspect, even though test_competitor is not designed to be used as reusable and extensible component, by changing all the private functions and variables in test_competitor to be protected level, it could be made somewhat extensible by means of inheritance. https://github.com/usnistgov/ARIAC/blob/ariac2023/test_competitor/include/test_competitor/test_competitor.hpp#L80

Of course, it's still possible to copy over the test_competitor class to teams solution and do away as they seem fit.

**Comment by 130s** [comment]
> by changing all the private functions and variables in test_competitor to be protected level, it could be made somewhat extensible by means of inheritance. https://github.com/usnistgov/ARIAC/blob/ariac2023/test_competitor/include/test_competitor/test_competitor.hpp#L80

I'm experimenting that https://github.com/usnistgov/ARIAC/pull/254

> Of course, it's still possible to copy over the test_competitor class to teams solution and do away as they seem fit.

Of course that's an option, and, 2023's submission setting doesn't seem to allow git clone-ing additional source repos (use of e.g. `vcstool`) as part of your application, bundling such codebases into yours might be inevitable. However, copy-pasting other git-managed codebase into your git codebase is just not scaling as a opensource community effort so IMO we should speak up to prevent that from happening. 

Hence I'm very +1 to this suggestion.
